zqd/space/space.go

```diff
@@ -34,59 +35,199 @@ var (
 	ErrSpaceNotExist       = zqe.E(zqe.NotFound, "space does not exist")
 )
 
+type Manager struct {
+	rootPath string
+	spaces   sync.Map
+}
+
 type Space struct {
 	path string
 	conf config
+
+	// state about operations in progress
+	opMutex       sync.Mutex
+	active        int
+	deletePending bool
+
+	wg sync.WaitGroup
+	// closed to signal non-delete ops should terminate
+	cancelChan chan struct{}
 }
 
-func Open(root, name string) (*Space, error) {
-	path := filepath.Join(root, name)
-	c, err := loadConfig(path)
+func NewManager(root string) *Manager {
+	mgr := &Manager{rootPath: root}
+
+	dirs, err := ioutil.ReadDir(root)
 	if err != nil {
-		if os.IsNotExist(err) {
-			return nil, ErrSpaceNotExist
+		// XXX log somewhere?
+		return mgr
+	}
+
+	for _, dir := range dirs {
+		if !dir.IsDir() {
+			continue
 		}
-		return nil, err
+
+		path := filepath.Join(root, dir.Name())
+		config, err := loadConfig(path)
+		if err != nil {
+			// XXX log?
+			continue
+		}
+
+		space := Space{
+			path:       path,
+			conf:       config,
+			cancelChan: make(chan struct{}, 0),
+		}
+		mgr.spaces.Store(space.Name(), &space)
 	}
-	return &Space{path, c}, nil
+
+	return mgr
 }
 
-func Create(root, name, dataPath string) (*Space, error) {
-	// XXX this should be validated before reaching here.
+func (m *Manager) Create(name, dataPath string) (api.SpacePostResponse, error) {
+	_, exists := m.spaces.Load(name)
+	if exists {
+		return api.SpacePostResponse{}, ErrSpaceExists
+	}
+
 	if name == "" && dataPath == "" {
-		return nil, errors.New("must supply non-empty name or dataPath")
+		return api.SpacePostResponse{}, errors.New("must supply non-empty name or dataPath")
```

Comment:
```suggestion
		return api.SpacePostResponse{}, zqe.E(zqe.Invalid, "must supply non-empty name or dataPath")
```
(I know the error was existing code)